R/tags.R

```diff
@@ -288,6 +291,94 @@ tagFunction <- function(func) {
   structure(func, class = "shiny.tag.function")
 }
 
+#' Add a tag render function
+#'
+#' Adds a render method to a tag object. This allows for the tag structure to
+#' be physically present while still allowing for a function to enhance (or
+#' even completely replace) the original tag object.
+#'
+#' It is recommended to use a render function over a [`tagFunction()`] whenever
+#' possible. By using a render method, a the tag structure is not a black box
+#' and can be inspected and altered before print time.
+#'
+#' Using [`tagFunction()`] is recommended if a stand-in tag structure does not
+#' make sense.
+#'
+#' @seealso [`tagFunction`]
+#' @param tag A tag object. See [`tag()`]
+#' @param func Function that accepts a single tag object.
+#' @param replace If `TRUE`, it will overwrite any previous render functions
+#' @return A [`tag()`] object with a `.render` field containing `func`.
+#'   When the returned tag is _rendered_ (such as with [`as.tags()`]),
+#'   this function will be called.
+#' @examples
+#' # Have a place holder div and return a span instead
+#' obj <- div("example", .render = function(x) {
+#'   x$name <- "span"
+#'   x
+#' })
+#' obj$name # "div"
+#' print(obj) # Prints as a `span`
+#'
+#' # Add a class to the tag
+#' # Should print a `span` with class `"extra"`
+#' spanExtra <- tagAddRenderFunction(obj, function(x) {
+#'   tagAppendAttributes(x, class = "extra")
+#' })
+#' spanExtra
+#'
+#' # Replace the previous render method
+#' # Should print a `div` with class `"extra"`
+#' divExtra <- tagAddRenderFunction(obj, replace = TRUE, function(x) {
+#'   tagAppendAttributes(x, class = "extra")
+#' })
+#' divExtra
+#'
+#' # Add more child tags
+#' spanExtended <- tagAddRenderFunction(obj, function(x) {
+#'   tagAppendChildren(x, " ", tags$strong("bold text"))
+#' })
+#' spanExtended
+#'
+#' # Add a new html dependency
+#' newDep <- tagAddRenderFunction(obj, function(x) {
+#'   fa <- htmlDependency(
+#'     "font-awesome", "4.5.0", c(href="shared/font-awesome"),
+#'     stylesheet = "css/font-awesome.min.css")
+#'   attachDependencies(x, fa, append = TRUE)
+#' })
+#' # Also add a jqueryui html dependency
+#' htmlDependencies(newDep) <- htmlDependency(
+#'   "jqueryui", "1.11.4", c(href="shared/jqueryui"),
+#'   script = "jquery-ui.min.js")
+#' # At render time, both dependencies will be found
+#' renderTags(newDep)$dependencies
+#'
+#' # Ignore the original tag and return something completely new.
+#' newObj <- tagAddRenderFunction(obj, function(x) {
+#'   tags$p("Something else")
+#' })
+#' newObj
+tagAddRenderFunction <- function(tag, func, replace = FALSE) {
+  if (!is.function(func) || length(formals(func)) == 0) {
+    stop("`func` must be a function that accepts at least 1 argument")
+  }
+
+  prevFunc <- tag$.render
+
+  if (!is.function(prevFunc) || isTRUE(replace)) {
+    tag$.render <- func
+    return(tag)
+  }
+
+  tag$.render <- function(x) {
+    func(prevFunc(x))
```

Comment:
Subtle issue here, which may be relevant. If I were a user, I'd expect the functions to be executed in the order that they're added, but because of R's lazy evaluation, this could cause them to execute in a different order, depending on the contents of the functions. This may be relevant in cases where the functions are altering and reading external state, which may be the case when (for example) setting the current theme.

For example, consider these functions:

```R
times_10 <- FALSE

first <- function(x) {
  times_10 <<- TRUE
  x
}

second <- function(x) {
  if (times_10) x * 10
  else              x
}
```

If you call `first()` and save the value, and then call `second()` on that value, you'd see this:

```R
times_10 <- FALSE
y <- first(1)
y <- second(y)
y
#> [1] 10
```

But if you call `second(first(1))`, you get a different answer:

```R
times_10 <- FALSE
y <- second(first(1))
y
#> [1] 1
```

So I think it would make sense to call the two functions separately, instead of nested.

Reply:
Oh, and I suppose there should be a test for this.